### PR TITLE
Update to serde 1.0

### PR DESCRIPTION
**Pull request by jeandudey**
**Status:** Ready for review

<hr/>

### CODE REVIEW

**Comment by apoelstra** [comment]
I think upgrading to serde 1.0 might be easier than trying to support both. I believe what needs to be done is:

1. Update `strason` to 0.4 and `secp256k1` to 0.10.0.
2. Put `strason`, `serde` and the whole `decimal` module under a feature gate. (Later we can change `decimal` to be compatible with `serde_json`s arbitrary_precision hack, drop the `strason` dependency, and make `serde_json` be a dev-dependency. But one step at a time.)
3. Replace the `serde_struct_impl` macro in src/internal_macros.rs. You'll find the hardest thing to replicate is that it tries to recover from a missing field by calling `MapVisitor::missing_field`. The modern way[*] to do this is to require `Default` be implemented on optional fields, and use `Default::default()` in place of `missing_field`. But as far as I can tell we never have optional fields in rust-bitcoin structs, so it's fine to just drop this "feature".

[*] Ok, yes, the actual modern way is to use `serde_derive`, but this macro lets us avoid the compiler version bump.

**Comment by jeandudey** [comment]
Also I've included your secp256k1 changes in the PR.

**Comment by apoelstra** [comment]
Can you rebase now that we have secp 0.10 in?

**Comment by jeandudey** [comment]
Done

File: src/lib.rs [hunk 1]
```diff
@@ -47,8 +47,8 @@ extern crate crypto;
 extern crate hex;
 extern crate rand;
 extern crate secp256k1;
-extern crate serde;
-extern crate strason;
+#[cfg(feature = "serde")] extern crate serde;
+#[cfg(any(feature = "serde-decimal", feature = "strason"))] extern crate strason;
```
Comment (apoelstra) [inline src/lib.rs]: I think this `any` isn't needed because `serde-decimal` implies `strason`?

**Comment by jeandudey** [comment]
Done

**Pull request merged**